### DISCUSSION TITLE
[8 febbraio 2024] Automi

### DISCUSSION THREAD

**Discussion by Antaarees** [sapienzastudentsnetwork/automi-calcolabilita-complessita#23] [Automi]
![20240213_113319](https://github.com/sapienzastudentsnetwork/acc2324/assets/19594279/602d3ea9-3b25-44c3-af6d-de8116065e3e)

**Comment by Exyss** [comment]
Per il primo punto, ripropongo le seguenti due soluzioni presenti nei miei appunti:

![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/65043207/2feb5e53-57f7-48ae-a8c8-7a92627b1d70)

----

Le due dimostrazioni richiesta nel secondo punto sono presenti sul libro di testo

**Reply by andrecaa** [reply]
Scusami ma nell'esempio di L2 lo stato finale non dovrebbe essere q1?

**Reply by Exyss** [reply]
Sì, hai ragione

**Reply by interestingmaneuver** [reply]
In realtà la seconda dimostrazione c'è come esercizio non risolto, dovrebbe essere così:

Una CFG in CNF ha regole esclusivamente della forma

* $A \to BC$
* $A \to a$

Per derivare una qualsiasi stringa $w$ di lunghezza $n$ devo arrivare a una derivazione che ha $n$ terminali. Inizio con la variabile iniziale $S$, ogni volta che applico le regole del primo tipo sostituisco una variabile con altre due, quindi se le applico $n - 1$ volte avrò $n$ variabili alla fine. Una volta che ho $n$ variabili, applico le regole del secondo tipo $n$ volte per mutare tutte le variabili in terminali, per un totale di $2n - 1$ passi.

Le regole del primo tipo sono l'unico modo per "allungare" la stringa derivata, e non c'è modo di "rimpicciolirla", perciò se il numero di passi fosse diverso da $2n - 1$ il risultato sarebbe sbagliato perché non avrebbe lunghezza $n$. Quindi in realtà ho anche dimostrato che $2n-1$ è il *lower bound* per i passi.

**Comment by CiottoloMaggico** [comment]
Propongo soluzione per il punto 2.
<img width="1216" height="406" alt="image" src="https://github.com/user-attachments/assets/ce53aeea-7a4f-454b-afc4-9bb49cae80aa" />

**Comment by daromacarlo** [comment]
**_Punto 1)_**
<img width="1273" height="686" alt="immagine" src="https://github.com/user-attachments/assets/dda556c1-04fb-4f4c-b7c0-17024ba7bb5c" />

_**Punto 2)**_
<img width="1131" height="512" alt="Immagine 2026-01-05 180325" src="https://github.com/user-attachments/assets/ca1692ab-0dea-40b2-8254-b3b44df3c7e6" />
<img width="1132" height="590" alt="Immagine 2026-01-05 180336" src="https://github.com/user-attachments/assets/7524d739-5617-4494-bee0-b9bf545d0604" />
